libmultilabel/nn/plt.py

```diff
@@ -0,0 +1,460 @@
+                weight_decay=self.config.weight_decay,
+                lr_scheduler=self.config.lr_scheduler,
+                scheduler_config=self.config.scheduler_config,
+                val_metric=self.config.val_metric,
+                metric_threshold=self.config.metric_threshold,
+                monitor_metrics=self.config.monitor_metrics,
+                multiclass=self.config.multiclass,
+                loss_function=self.config.loss_function,
+                silent=self.config.silent,
+                save_k_predictions=self.config.save_k_predictions,
+            )
+
+            logger.info(f"Training level 0. Number of labels: {num_labels}")
+            trainer.fit(model, train_dataloader, val_dataloader)
+            logger.info(f"Finish training level 0")
+
+        logger.info(f"Best model loaded from {best_model_path}")
+        model = Model.load_from_checkpoint(best_model_path, embed_vecs=self.embed_vecs)
+
+        # Utilize single GPU to predict
+        trainer = Trainer(
+            num_nodes=1,
+            devices=1,
+            accelerator=self.accelerator,
+            logger=False,
+        )
+        logger.info(
+            f"Generating predictions for level 1. Number of possible predictions: {num_labels}. Top k: {self.predict_top_k}"
+        )
+        # load training and validation data and predict corresponding level 0 nodes
+        train_dataloader = self.eval_dataloader(MultiLabelDataset(train_x))
+        val_dataloader = self.eval_dataloader(MultiLabelDataset(val_x))
+
+        train_node_pred = trainer.predict(model, train_dataloader)
+        valid_node_pred = trainer.predict(model, val_dataloader)
+
+        # shape of node_pred: (n, 2, ~batch_size, top_k). n is floor(num_x / batch_size)
+        # new shape: (2, num_x, top_k)
+        _, train_node_y_pred = map(torch.vstack, list(zip(*train_node_pred)))
+        valid_node_score_pred, valid_node_y_pred = map(torch.vstack, list(zip(*valid_node_pred)))
+
+        logger.info("Selecting relevant/irrelevant labels of each instance for level 1 training")
+        cluster_samples = np.empty((len(train_x), self.predict_top_k), dtype=np.int64)
+        for i, ys in enumerate(tqdm(train_node_y_pred, leave=False, desc="Sampling")):
+            # true nodes/labels are positive
+            pos = set(train_y_clustered.indices[train_y_clustered.indptr[i] : train_y_clustered.indptr[i + 1]])
+            # Regard positive nodes and predicted training nodes that are not in positive as samples
+            # until reaching top_k if the number of positive labels is less than top_k.
+            if len(pos) <= self.predict_top_k:
+                samples = pos
+                for y in ys:
+                    y = y.item()
+                    if len(samples) == self.predict_top_k:
+                        break
+                    samples.add(y)
+            # Regard positive (true) label as samples iff they appear in the predicted labels
```

Comment:
How about

As number of relevant labels is more than to be selected, from top-predicted clusters we sequentially include relevant labels in them until the total number reaches ??.

Reply:
what's the issue of this comment?